## Практическая работа

В этой практической работе пять обязательных задач.

*Обязательные задачи* помогут проверить, что вы действительно усвоили материал модуля. Сдайте их на проверку.

Удачи!

Цели практической работы: 


1.   Потренироваться в обучении модели линейной регрессии.
2.   Потренироваться в обучении модели логистической регрессии. 
3.   Потренироваться в обучении модели многослойного персептрона.





Что оценивается:

*   Все пункты и критерии приёмки по каждому заданию выполнены.
*   Код оформлен и работает корректно.





Как отправить работу на проверку:

Сдайте практическую работу этого модуля через систему контроля версий Git сервиса Skillbox GitLab. После загрузки работы на проверку напишите об этом в личном кабинете своему куратору.

В этом задании вам нужно построить модель классификации, определяющую категорию цены/цену подержанного автомобиля в зависимости от характеристик транспортного средства. 

Мы будем использовать датасет из коллекции подержанных автомобилей, выставленных на продажу в Соединенных Штатах. Он уже подготовлен, без выбросов и с категориальными фичами, преобразованными с помощью one hot encoding, и количественными фичами, стандартизированными с помощью скейлеров.


### Описание датасета:
- `id`: идентификатор записи;
- `is_manufacturer_name`: признак производителя автомобиля;

- `region_*`: регион;
- `x0_*`: тип топлива;
- `manufacturer_*`: производитель;
- `short_model_*`: сокращённая модель автомобиля;
- `title_status_*`: статус;
- `transmission_*`: коробка передач;
- `state_*`: штат;
- `age_category_*`: возрастная категория автомобиля;

- `std_scaled_odometer`: количество пройденных миль (после стандартизации);
- `year_std`: год выпуска (после стандартизации);
- `lat_std`: широта (после стандартизации);
- `long_std`: долгота (после стандартизации);
- `odometer/price_std`: отношение стоимости к пробегу автомобиля (после стандартизации);
- `desc_len_std`: количество символов в тексте объявления о продаже (после стандартизации);
- `model_in_desc_std`: количество наименований модели автомобиля в тексте объявления о продаже (после стандартизации);
- `model_len_std`: длина наименования автомобиля (после стандартизации);
- `model_word_count_std`: количество слов в наименовании автомобиля (после стандартизации);
- `month_std`: номер месяца размещения объявления о продаже автомобиля (после стандартизации);
- `dayofweek_std`: день недели размещения объявления о продаже автомобиля (после стандартизации);
- `diff_years_std`: количество лет между годом производства автомобиля и годом размещения объявления о продаже автомобиля (после стандартизации);

- `price`: стоимость;
- `price_category`: категория цены.

## Обязательные задачи

In [1]:
%%time
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix, mean_absolute_error
from sklearn.model_selection import train_test_split 
from sklearn.neural_network import MLPClassifier

Wall time: 228 ms


In [2]:
%%time
df = pd.read_csv('data/vehicles_dataset_prepared.csv')
df.head()

Wall time: 1.91 s


id  price price_category  is_audi  is_ford  is_chevrolet  \
0  7308295377  54990           high        0        0             0   
1  7316380095  16942         medium        0        1             0   
2  7313733749  35590           high        0        0             0   
3  7308210929  14500         medium        0        0             0   
4  7303797340  14590         medium        0        0             0   

   is_toyota  x0_diesel  x0_electric  x0_gas  ...  long_std  year_std  \
0          0        1.0          0.0     0.0  ...  0.484245  1.322394   
1          0        0.0          0.0     0.0  ...  1.110800  0.695973   
2          0        0.0          0.0     1.0  ...  0.531185  0.852578   
3          1        0.0          0.0     1.0  ...  0.853562  0.226157   
4          0        0.0          0.0     0.0  ...  0.557607  0.069552   

   odometer/price_std  desc_len_std  model_in_desc_std  model_len_std  \
0           -0.510784      0.632075          -0.155788       1.163032   
1           -0.402947     -0.646781          -0.155788       0.932087   
2           -0.514480      0.560744          -0.155788       0.470197   
3           -0.241883      0.180435          -0.155788      -0.915473   
4           -0.333074      0.766366          -0.155788       1.163032   

   model_word_count_std  month_std  dayofweek_std  diff_years_std  
0              1.910669  -0.615846       1.120284       -1.322394  
1              1.235799   1.623784      -1.374972       -0.695973  
2              0.560930  -0.615846      -0.376870       -0.852578  
3             -0.788810  -0.615846       1.120284       -0.226157  
4              1.910669  -0.615846       0.122182       -0.069552  

[5 rows x 1463 columns]

**Задача 0. Линейная регрессия**

Вспомните задачу по предсказанию стоимости поддержанного автомобиля. Попробуем обучить модель линейной регрессии для предсказания цены автомобиля (колонка `price`). Для этого проделайте следующие шаги:

- Подготовьте данные: удалите колонки, которые косвенно содержат информацию о целевой переменной (`odometer/price_std`, `price_category`).
- Разделите выборку на треин и тест в отношении 70/30.
- Обучите модель линейной регрессии с дефолтными параметрами.
- Посчитайте значение метрики mae на тестовой выборке для линейной регрессии.
- Выведите получившиеся коэффициенты линейной регрессии при каждом параметре обучающей выборки с помощью метода `coef_`. Есть ли коэффициенты, которые равны нулю? Если есть, выведите названия фичей с нулевым коэффициентом
- Удалите фичи, коэффициенты которых равны нулю. И переобучите модель. Убедитесь, что значение метрики не изменилось.



In [3]:
%%time
# Подготовка данных: удалим переменные, которые косвенно содержат информацию о целевой переменной и тип данных str
print("Строковые данные", df.select_dtypes(include='object').columns.values)
df_prepared= df.drop(['odometer/price_std', 'price_category'], axis=1)
df.shape, df_prepared.shape

Строковые данные ['price_category']
Wall time: 33 ms


((9619, 1463), (9619, 1461))

In [4]:
%%time
# Разделим выборку на треин и тест в отношении 70/30
x = df_prepared.drop(['price'], axis=1)
y = df_prepared['price']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)
x_train.shape, x_test.shape

Wall time: 119 ms


((6733, 1460), (2886, 1460))

In [5]:
%%time
# Обучим модель линейной регрессии с дефолтными параметрами.
#Создадим экземпляр класса Линейной регрессии
reg =  LinearRegression()
reg.fit(x_train, y_train)
predicted_test = reg.predict(x_test)
#Посчитайте значение метрики mae на тестовой выборке для линейной регрессии
mae1 = mean_absolute_error(y_test, predicted_test)
print("Значение метрики MAE на тестовой выборке = ", mae1)

Значение метрики MAE на тестовой выборке =  4600.341116727147
Wall time: 1.25 s


In [6]:
%%time
# Выведим получившиеся коэффициенты линейной регрессии при каждом параметре обучающей выборки с помощью метода coef_
cdf = pd.DataFrame(reg.coef_, x_train.columns, columns=['Coeff'])
cdf.head()

Wall time: 1e+03 µs


Coeff
id               0.000006
is_audi        254.298285
is_ford       1073.906344
is_chevrolet  -184.762537
is_toyota      471.497678

In [7]:
%%time
# Коэффициенты, которые равны нулю
print(cdf[cdf.Coeff ==0].index.values)
#Удалим переменную с нулевым весом
x_train, x_test = x_train.drop(cdf[cdf.Coeff ==0].index.values, axis=1), x_test.drop(cdf[cdf.Coeff ==0].index.values, axis=1)
x_train.shape, x_test.shape

['manufacturer_ferrari']
Wall time: 40 ms


((6733, 1459), (2886, 1459))

In [8]:
%%time
# Переобучим модель и посчитаем метрику
reg.fit(x_train, y_train)
predicted_test = reg.predict(x_test)
#Посчитайте значение метрики mae на тестовой выборке для линейной регрессии
mae2 = mean_absolute_error(y_test, predicted_test)
print("Значение метрики MAE на тестовой выборке = ", mae2)
print("Разница оценок = ", mae2-mae1)

Значение метрики MAE на тестовой выборке =  4600.341116175135
Разница оценок =  -5.520123522728682e-07
Wall time: 1.36 s


In [9]:
import time
def exec_time(start, end):
    diff_time = end - start

**Задача 1. Логистическая регрессия**

Теперь в рамках тех же данных попробуйте предсказать `price_category` с помощью алгоритма логистической регрессии. Предварительно из датафрейма удалите переменные, в которых косвенно содержится информация о целевой переменной (`odometer/price_std`, `price`). 

Для обученной модели нужно:

- рассчитать и вывести метрику качества (accuracy) на тренировочной выборке;
- сделать предикт на тестовых данных и положить его в переменную `logreg_pred`;
- рассчитать и вывести accuracy и confusion_matrix на тестовой выборке.


Задание засчитывается, если: 
- accuracy на тренировочной выборке > 87%;
- accuracy на тестовой выборке > 75.5%.

Для улучшения метрики можно преобразовывать входные данные, тюнить параметры логистической регрессии.

In [10]:
%%time
# Ваш код подготовки данных здесь. Допишите инициализацию входных фичей и целевую переменную
df_prepared= df.drop(['odometer/price_std', 'price'], axis=1)
x = df_prepared.drop(['price_category'], axis=1)
y = df_prepared['price_category']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)
x_train.shape, x_test.shape

Wall time: 207 ms


((6733, 1460), (2886, 1460))

In [11]:
%%time
# Обучим модель логистической регрессии с дефолтными параметрами.
# Создадим экземпляр класса Линейной регрессии
logreg =  LogisticRegression()
logreg.fit(x_train, y_train)
predicted_logreg_train = logreg.predict(x_train)
predicted_logreg_test = logreg.predict(x_test)
#Посчитайте значение метрики mae на тестовой выборке для линейной регрессии
acc_train = accuracy_score(y_train, predicted_logreg_train)
acc_test = accuracy_score(y_test, predicted_logreg_test)
print("Значение метрики accuracy на тренировочной выборке = ", acc_train,'\n',"Значение метрики accuracy на тестовой выборке = ", acc_test )

Значение метрики accuracy на тренировочной выборке =  0.3392247140947572 
 Значение метрики accuracy на тестовой выборке =  0.32432432432432434
Wall time: 446 ms


In [12]:
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
conf_matrix_baseline = pd.DataFrame(confusion_matrix(y_test, predicted_logreg_test), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix_baseline)
print('Модель логистической регрессии с дефолтными параметрами accuracy_score = ', acc_test)

predicted 0  predicted 1  predicted 2
high              0            0          997
medium            0            0          953
low               0            0          936

Модель логистической регрессии с дефолтными параметрами accuracy_score =  0.32432432432432434


In [13]:
%%time
logreg.get_params()

Wall time: 0 ns


{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'auto',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [14]:
%%time
#Удалим из данных переменную ID
x_train, x_test = x_train.drop(['id'], axis=1), x_test.drop(['id'], axis=1)
print(x_train.shape, x_test.shape)
#Обучим модель снова и посчитаем метрику
logreg.fit(x_train, y_train)
predicted_logreg_train = logreg.predict(x_train)
predicted_logreg_test = logreg.predict(x_test)
#Посчитайте значение метрики mae на тестовой выборке для линейной регрессии
acc_train = accuracy_score(y_train, predicted_logreg_train)
acc_test = accuracy_score(y_test, predicted_logreg_test)
print("Значение метрики accuracy на тренировочной выборке = ", acc_train,'\n',"Значение метрики accuracy на тестовой выборке = ", acc_test )

(6733, 1459) (2886, 1459)
Значение метрики accuracy на тренировочной выборке =  0.8551908510322294 
 Значение метрики accuracy на тестовой выборке =  0.7543312543312544
Wall time: 1.64 s


D:\Users\garri\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [15]:
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
conf_matrix_baseline = pd.DataFrame(confusion_matrix(y_test, predicted_logreg_test), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix_baseline)
print('Модель логистической регрессии с дефолтными параметрами без переменной ID accuracy_score = ', acc_test)

predicted 0  predicted 1  predicted 2
high            801           26          170
medium           22          767          164
low             156          171          609

Модель логистической регрессии с дефолтными параметрами без переменной ID accuracy_score =  0.7543312543312544


In [16]:
%%time
# Увеличим количество итераций (max_iter) следуя рекомендации
logreg_opt = LogisticRegression(max_iter=200,n_jobs=-1)
#Обучим модель снова и посчитаем метрику
logreg_opt.fit(x_train, y_train)
predicted_logreg_train = logreg_opt.predict(x_train)
predicted_logreg_test = logreg_opt.predict(x_test)
#Посчитайте значение метрики mae на тестовой выборке для линейной регрессии
acc_train = accuracy_score(y_train, predicted_logreg_train)
acc_test = accuracy_score(y_test, predicted_logreg_test)
print("Значение метрики accuracy на тренировочной выборке = ", acc_train,'\n',"Значение метрики accuracy на тестовой выборке = ", acc_test )

Значение метрики accuracy на тренировочной выборке =  0.8554878954403683 
 Значение метрики accuracy на тестовой выборке =  0.7571032571032571
Wall time: 6.82 s


In [17]:
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
conf_matrix_baseline = pd.DataFrame(confusion_matrix(y_test, predicted_logreg_test), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix_baseline)
print('Модель логистической регрессии с дефолтными параметрами без переменной ID с увеличенным количество итераций accuracy_score = ', acc_test)

predicted 0  predicted 1  predicted 2
high            803           26          168
medium           22          772          159
low             155          171          610

Модель логистической регрессии с дефолтными параметрами без переменной ID с увеличенным количество итераций accuracy_score =  0.7571032571032571


In [18]:
%%time
# Оптимизируем параметры модели для достижения поставленной цели accuracy на тренировочной выборке > 87%, на тестовой выборке > 75.5%.
logreg_opt = LogisticRegression(max_iter=200,n_jobs=-1,C=2)
#Обучим модель снова и посчитаем метрику
logreg_opt.fit(x_train, y_train)
predicted_logreg_train = logreg_opt.predict(x_train)
predicted_logreg_test = logreg_opt.predict(x_test)
#logreg_pred = predicted_logreg_test
#Посчитайте значение метрики mae на тестовой выборке для линейной регрессии
acc_train = accuracy_score(y_train, predicted_logreg_train)
acc_test = accuracy_score(y_test, predicted_logreg_test)
print("Значение метрики accuracy на тренировочной выборке = ", acc_train,'\n',"Значение метрики accuracy на тестовой выборке = ", acc_test )

Значение метрики accuracy на тренировочной выборке =  0.8706371602554582 
 Значение метрики accuracy на тестовой выборке =  0.7584892584892585
Wall time: 6.1 s


In [19]:
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
conf_matrix_baseline = pd.DataFrame(confusion_matrix(y_test, predicted_logreg_test), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix_baseline)
print('Модель логистической регрессии с дефолтными параметрами без переменной ID с увеличенным количество итераций и увеличенным параметром С (Обратная сила регуляризации)','\n','accuracy_score = ', acc_test)

predicted 0  predicted 1  predicted 2
high            802           22          173
medium           22          772          159
low             156          165          615

Модель логистической регрессии с дефолтными параметрами без переменной ID с увеличенным количество итераций и увеличенным параметром С (Обратная сила регуляризации) 
 accuracy_score =  0.7584892584892585


**Задача 2. Многослойный персептрон**

Решите задачу, поставленную в предыдущем задании, применив модель многослойного персептрона. Добейтесь лучших значений точности на тренировочной и тестовой выборках, чем на логистической регрессии, с помощью данного алгоритма. 

Значимым увеличением точности в данном задании будем считать увеличение больше, чем на 0.01 по метрике accuracy.

In [20]:
%%time
# Модель многослойного персептрона
mlp = MLPClassifier(random_state=42,
                    max_iter=500,
                    hidden_layer_sizes=(100,20),
                    activation = 'tanh')
mlp.fit(x_train, y_train)

Wall time: 1min 42s


MLPClassifier(activation='tanh', hidden_layer_sizes=(100, 20), max_iter=500,
              random_state=42)

In [21]:
%%time
predict_mlp_train = mlp.predict(x_train)
predict_mlp_test = mlp.predict(x_test)
#Посчитайте значение метрики mae на тестовой выборке для Многослойного персептрона
acc_train = accuracy_score(y_train, predict_mlp_train)
acc_test = accuracy_score(y_test, predict_mlp_test)
print("Значение метрики accuracy на тренировочной выборке = ", acc_train,'\n',"Значение метрики accuracy на тестовой выборке = ", acc_test )

Значение метрики accuracy на тренировочной выборке =  1.0 
 Значение метрики accuracy на тестовой выборке =  0.7761607761607762
Wall time: 174 ms


In [22]:
%%time
mlp.get_params()

Wall time: 0 ns


{'activation': 'tanh',
 'alpha': 0.0001,
 'batch_size': 'auto',
 'beta_1': 0.9,
 'beta_2': 0.999,
 'early_stopping': False,
 'epsilon': 1e-08,
 'hidden_layer_sizes': (100, 20),
 'learning_rate': 'constant',
 'learning_rate_init': 0.001,
 'max_fun': 15000,
 'max_iter': 500,
 'momentum': 0.9,
 'n_iter_no_change': 10,
 'nesterovs_momentum': True,
 'power_t': 0.5,
 'random_state': 42,
 'shuffle': True,
 'solver': 'adam',
 'tol': 0.0001,
 'validation_fraction': 0.1,
 'verbose': False,
 'warm_start': False}

In [23]:
%%time
mlp.n_layers_

Wall time: 0 ns


4

In [24]:
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
conf_matrix_baseline = pd.DataFrame(confusion_matrix(y_test, predict_mlp_test), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix_baseline)
print('Модель многослойного персептрона без переменной ID accuracy_score =', acc_test)

predicted 0  predicted 1  predicted 2
high            813           34          150
medium           18          763          172
low             125          147          664

Модель многослойного персептрона без переменной ID accuracy_score = 0.7761607761607762


**Задача 3. Сравнение с древовидными моделями**

Обучите модель случайного леса на тех же данных для предсказания `price_category`. Сравните качество с моделью логистической регрессии и многослойного персептрона. Словами опишите, какая из моделей в каких случаях работает лучше по результатам на тестовой выборке, и обоснуйте свой выбор.

In [25]:
%%time
# Импортируем модуль
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_jobs=-1,
                            random_state=42,
                            n_estimators = 700,
                            max_features='log2',
                            max_depth=120,
                            bootstrap=False)
rfc.fit(x_train, y_train)

Wall time: 5.91 s


RandomForestClassifier(bootstrap=False, max_depth=120, max_features='log2',
                       n_estimators=700, n_jobs=-1, random_state=42)

In [26]:
%%time
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
predict_rfc_test = rfc.predict(x_test)
conf_matrix = pd.DataFrame(confusion_matrix(y_test, predict_rfc_test), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix)
print('Random Forest RandomizedSearchCV accuracy_score = ', accuracy_score(y_test, predict_rfc_test))

predicted 0  predicted 1  predicted 2
high            812           36          149
medium           15          814          124
low             133          202          601

Random Forest RandomizedSearchCV accuracy_score =  0.7716562716562717
Wall time: 402 ms


***Точность моделей:***
    1. Многослойного персептрона = 77.61%
    2. Случайный лес = 77.16%
    3. Логистической регрессии = 75.84%

***Время затраченное на обучение модели:***
    1. Случайный лес = 5.9s
    2. логистической регрессии = 6.1s,
    3. Многослойного персептрона = 102.0s,

Самой точной оказалась модель Многослойного персептрона, но она же заняла больше всего времени на обучение на не большом датасете 9619 строк и 1461 перемаенная. По результату сравнительного анализа качества и времени предпочтение отдано модели случайныйного леса, у этой модели качество не значительно ниже Многослойного персептрона ****77,2%**** против ****77,6%**** соответственно, но скорость обучения в ****20**** раз выше чем у конкурента.

***Распределение моделей по метрике (скорость обучения - качество):***
1. Случайный лес
2. Многослойный персептрон
3. Логистическая регрессия
Логистическую регрессию можно использовать в качестве базовой модели (Base Line)

**Задача 4. Стратегия голосования**

Реализуйте стратегию голосования для предсказания целевой переменной.
Голосование в задаче классификации — это когда несколько моделей выдают свои предикты, и финальным выбирается тот предикт, который предсказали большинство моделей.

Для реализации этой стратегии проделайте следующее:

- сохраните предсказания каждой из моделей (случайный лес, многослойный персептрон, логистическая регрессия) для тестовой выборки в датафрейм `pred_df`;
- в четвертую колонку `target` положите тот класс, который предсказало большинство классификаторов. Например, если в строке были значения `high, medium, medium`, в `target` нужно положить `medium`.

     Если в строке три разных класса — `high, medium, low`, — придумайте свою стратегию по выбору значения. Самая простая стратегия: выбрать рандомно одно значение из трёх.

- посчитайте точность предсказания с помощью голосования. Выведите значения метрик accuracy и confusion_matrix.


Добейтесь значения точности > 78%

In [27]:
%%time
# Функция голосования для предсказания целевой переменной
import random
def target_pred(x):
    if  x[0]==x[1]:
        return x[0]
    elif x[1]==x[2]:
        return x[1]
    elif  x[2]==x[0]:
        return x[2]
    else:
        return random.choice([x[0], x[1], x[2]])

Wall time: 0 ns


In [28]:
%%time
# Создание DataFrame и заполнение переменной target
pred_df = pd.DataFrame({'predicted_logreg_test':predicted_logreg_test,'predict_mlp_test':predict_mlp_test,'predict_rfc_test':predict_rfc_test})
pred_df['target'] = pred_df.apply(target_pred, axis=1)
pred_all = pred_df.target
#Создадим матрицы ошибок для моделей и узнаем о том, как хорошо каждая из них способна предсказывать
conf_matrix = pd.DataFrame(confusion_matrix(y_test, pred_all), index = ['high', 'medium', 'low'], columns = ['predicted 0', 'predicted 1', 'predicted 2'])
display(conf_matrix)
print('Голосование предсказаний 3-ч моделей в задаче классификации accuracy_score = ', accuracy_score(y_test, pred_all))

predicted 0  predicted 1  predicted 2
high            816           27          154
medium           16          806          131
low             135          166          635

Голосование предсказаний 3-ч моделей в задаче классификации accuracy_score =  0.782051282051282
Wall time: 60.5 ms


**Примечание:**

В этом домашнем задании в задаче номер 1 и далее перед вами встал выбор: включать ли в обучение колонку `id`. При удалении данной колонки во время обучении логистической регрессии качество заметно улучшается.

Вообще обучать любую модель (будь то логистическая или линейная регрессия, или древовидный алгоритм) на данных айдишников не считается хорошей практикой. Как правило, между `id` и целевой переменной не должно быть никаких взаимосвязей. Включая колонку `id` в качества атрибута в обучение, вы стараетесь "заточить" результаты своей модели под айдишники записей. Таким образом модель обучится на некотором наборе "частных случаев", и, возможно, не обратит внимание на общие зависимости.